mart/attack/threat_model.py

```diff
@@ -4,43 +4,122 @@
 # SPDX-License-Identifier: BSD-3-Clause
 #
 
+from __future__ import annotations
+
 import abc
-from typing import Any, Dict, Union
+from typing import Any
 
 import torch
 
-__all__ = ["BatchThreatModel"]
+__all__ = ["BatchComposer"]
+
+
+class Constraint(abc.ABC):
+    @abc.abstractclassmethod
+    def __call__(self, input, target, input_adv) -> None:
```

Comment:
`input_adv` is moved to the first place.

I am not making `input` and `target` optional because all constraints are fed with the same arguments in a for loop.

Reply:
Thanks for the suggestion. `input` and `target` are made keyword arguments.